web/packages/teleterm/src/services/appUpdater/autoUpdatesStatus.ts

```diff
@@ -0,0 +1,255 @@
+  clusterVersions: GetClusterVersionsResponse;
+}): AutoUpdatesStatus {
+  const { reachableClusters, unreachableClusters } = sources.clusterVersions;
+  const candidateClusters = makeCandidateClusters(reachableClusters);
+  if (!candidateClusters.length) {
+    return {
+      enabled: false,
+      reason: 'no-cluster-with-auto-update',
+      unreachableClusters,
+    };
+  }
+
+  const clusterUri = sources.managingClusterUri;
+  if (clusterUri) {
+    const cluster = candidateClusters.find(c => c.clusterUri === clusterUri);
+    if (cluster) {
+      return {
+        enabled: true,
+        version: cluster.toolsVersion,
+        source: {
+          kind: 'managing-cluster',
+          candidateClusters,
+          clusterUri,
+          unreachableClusters,
+        },
+      };
+    }
+    logger.warn(
+      `Cluster ${clusterUri} managing updates was not found, continuing resolution.`
+    );
+  }
+
+  const mostCompatibleVersion =
+    findMostCompatibleToolsVersion(candidateClusters);
+  if (mostCompatibleVersion) {
+    return {
+      enabled: true,
+      version: mostCompatibleVersion,
+      source: {
+        kind: 'most-compatible',
+        candidateClusters,
+        unreachableClusters,
+        clustersUri: candidateClusters
+          .filter(c => c.toolsVersion === mostCompatibleVersion)
+          .map(c => c.clusterUri),
+      },
+    };
+  }
+
+  return {
+    enabled: false,
+    reason: 'no-compatible-version',
+    unreachableClusters,
+    candidateClusters,
+  };
+}
+
+/** Assigns each candidate a compatibility with client tools from other clusters. */
+function makeCandidateClusters(
+  versions: ClusterVersionInfo[]
+): CandidateCluster[] {
+  const candidates = versions.filter(v => v.toolsAutoUpdate);
+
+  return candidates.map(candidate => {
+    const candidateMajorToolsVersion = getMajor(candidate.toolsVersion);
+
+    const otherCompatibleClusters = candidates
+      .filter(v => {
+        // The list should only contain other clusters.
+        if (v.clusterUri === candidate.clusterUri) {
+          return false;
+        }
+
+        const minMajor = getMajor(v.minToolsVersion);
+        const maxMajor = getMajor(v.toolsVersion);
+
+        // Check compatibility: candidate's major version must be within [minMajor, maxMajor].
+        return (
+          candidateMajorToolsVersion >= minMajor &&
+          candidateMajorToolsVersion <= maxMajor
+        );
+      })
+      .map(v => v.clusterUri);
+
+    return {
+      ...candidate,
+      otherCompatibleClusters,
+    };
+  });
+}
+
+/** Finds version of client tools that is compatible with all other clusters. */
+function findMostCompatibleToolsVersion(
+  candidates: CandidateCluster[]
+): string | undefined {
+  // Get the highest version first.
+  const sorted = candidates.toSorted((a, b) =>
+    compareSemVers(b.toolsVersion, a.toolsVersion)
+  );
+
+  const requiredClusters = new Set(candidates.map(c => c.clusterUri));
```

Comment:
What does "required" mean in this context? This adjective is not used outside of this function. Should it be something more along the lines of "allCandidates"?